Added connection distribution check.

J-Gras opened the pull request:
I have added a connection distribution check based on the relative standard deviation. Unfortunately, I have only limited experience operating larger clusters. Thus, I am neither sure about the threshold nor the approach itself.

One thing to definitively keep in mind is that people might feed different traffic streams to different workers on purpose. In that case an "uneven" distribution of traffic is intended.

So, everything up for discussion :)

J-Gras committed:
Added connection distribution check.

J-Gras committed:
Added check_connection_distribution to README.

JustinAzoff commented on doctor.py:
```diff
@@ -346,6 +347,43 @@ def check_duplicate_5_tuples(self):
+
+        nodes = defaultdict(int)
+        for rec in read_bro_logs_with_line_limit(reversed(files), 10000):
+            # Only count connections that have completed a three way handshake
```
Probably want to remove the if statements here.. those are mostly useful for checking for missed_bytes and for SAD/had connections.  For testing load balancing, would likely want to look at all connections regardless of if they look broken or not.

JustinAzoff commented on doctor.py:
```diff
@@ -346,6 +347,43 @@ def check_duplicate_5_tuples(self):
+        variance = reduce(lambda var, cnt: var + (cnt - mean)**2, nodes.values(), 0) / len(nodes)
+        rsd = sqrt(variance) / mean
+
+        if rsd > 0.1:
```
Should we have a 

    if len(nodes) == 1:
        self.ok("Only one worker appears to be in use, unable to check distribution")

JustinAzoff reviewed (commented):
Awesome!

Had one note about removing the if blocks, and another about what should be output in the case where there is only one worker in use.

J-Gras committed:
Improved connection distribution check.

Flipped connections and connections without data are now considered as
well. In case of only one node, the check is skipped.

JustinAzoff commented:
Now that I think about this more, should maybe be something like

    node = rec.get('_node_name', 'no-node-names')

and then use that info display a message about installing the node names bro-pkg.  Right now for our cluster it outputs:

This way it can differentiate between a standalone single 'bro' worker, and an actual cluster that's just missing the node names plugin.

It's also possible that someone has my older script installed, like this one: https://github.com/michalpurzynski/bro-gramming/blob/master/conn-peer.bro  that adds the node name to just the conn log as the 'peer' column.. so maybe there should be a helper function that takes a conn log record and tries to return either` _node_name` or `peer`.

J-Gras commented:
Good idea! Another candidate would be `node` (see https://github.com/0xxon/bro-scripts/blob/master/conn-workers.bro).

JustinAzoff commented:
Ah, good catch... so something like this?

```
NODE_KEYS = "_node_name", "node", "peer"
def get_node_name(rec):
    for key in NODE_KEYS:
        node = rec.get(key)
        if node:
            return node
```

Could probably optimize it to cache the last found key, but it's only going to be called 100k times or so so shouldn't matter too much.

JustinAzoff commented:
Oh!  I just thought of something else too.. can probably add it in another change though.

There's two levels of load balancing, the host level load balancing, and then the process level.  Using the bro config info one can turn a node name into a host and use that to count connections per host as well.

J-Gras committed:
Consider different keys for node names.

J-Gras commented:
I added support for multiple node name keys. One thing one could add is a warning if the number of packets is exactly the same on all nodes.

JustinAzoff commented:
Ah, that's a good idea, I think the duplicate connection check will cover the causes of that though... if all interfaces are all receiving the same traffic, that check will flag 100% of conns as duplicate.

JustinAzoff commented:
Looks good to me!  I'll see about having it also check machine-level distribution.
